### PR TITLE
Speed up progressive rendering

### LINKED ISSUE #395: Progressive rendering is super slow
Since progressive rendering has been added, rendering a Notebook in voila became super slow...

Taking a Notebook that prints "hello" 30 times:
note

I assume we all agree this Notebook should take less than a second to be executed, starting time of the kernel included.

Without the progressive rendering (using a dev version that simply removes this feature):
without

With the progressive rendering:
with

We should either improve this feature, or make it optional

### PR DESCRIPTION
**Pull request by martinRenou**
Should fix #395 

Instead of using the `preprocess` method (which is for preprocessing an entire Notebook) for each cell, I use `preprocess_cell` for each cell

speedup

### CODE REVIEW

**Comment by martinRenou** [comment]
@maartenbreddels what do you think? It improves perfomances but the logic was also wrong before (processing each cell like it's a single Notebook)

**Comment by martinRenou** [comment]
Part of the slowness was coming from the fact that internally nbconvert creates new sockets for connecting with the kernel at each `preprocess` call (that is to say for each cell...)

**Comment by SylvainCorlay** [comment]
Reviewed this in person with @martinRenou and this looks great.

**Comment by martinRenou** [comment]
I'm wondering if the `preprocess` implemented in the Voila executor has extra logic that we should have. Maybe the Voila executor should reimplement `preprocess_cell`

**Comment by martinRenou** [comment]
I now reimplement `preprocess_cell` in the Voila executor so that it has a similar behavior to `preprocess`

File: voila/handler.py [hunk 1]
```diff
@@ -133,15 +135,14 @@ def _jinja_cell_generator(self, nb, kernel_id):
...
-        for cell in all_cells:
-            # we execute one cell at a time
-            nb.cells = [cell]  # reuse the same notebook
-            result = executenb(nb, km=km, cwd=self.cwd, config=self.traitlet_config)
-            cell = result.cells[0]  # keep a reference to the executed cell
-            nb.cells = all_cells  # restore notebook in case we access it from the template
-            # we don't filter empty cells, since we do not know how many empty code cells we will have
-            yield cell
+        nb, resources = ClearOutputPreprocessor().preprocess(nb, {'metadata': {'path': self.cwd}})
+        ep = VoilaExecutePreprocessor(config=self.traitlet_config)
```
Comment (maartenbreddels) [inline voila/handler.py]: We are missing https://github.com/voila-dashboards/voila/commit/c21b216e227656521da419a82815f001e510b4b0  (see https://github.com/voila-dashboards/voila/blame/master/voila/execute.py#L159)

Comment (martinRenou) [reply voila/handler.py]: Mmh... Isn't it line `138`?

File: voila/execute.py [hunk 2]
```diff
@@ -107,6 +107,14 @@ def preprocess(self, nb, resources, km=None):
...
         return result
 
+    def preprocess_cell(self, cell, resources, cell_index, store_history=True):
+        try:
+            result = super(VoilaExecutePreprocessor, self).preprocess_cell(cell, resources, cell_index, store_history)
+        except CellExecutionError as e:
+            self.log.error(e)
+            result = [cell]
+        return result
+
```
Comment (maartenbreddels) [inline voila/execute.py]: Why is this needed?

Comment (martinRenou) [reply voila/execute.py]: To mimic the `VoilaExecutePreprocessor.preprocess` behavior. I change this behavior a bit in a separate PR concerning error outputs BTW.

Comment (martinRenou) [reply voila/execute.py]: The `try except` prevents a `505` server error when the Notebook contains an error
